### PR TITLE
fixes req.URL.Path for unix: sockets

### PR DESCRIPTION
**Pull request by eiszfuchs**
Please review my change like last time; I haven't been able to achieve master skills of Go yet.

# What's currently going wrong

Given a Caddyfile like this:

```text
127.0.0.1:2015
tls off

proxy / unix:/var/run/www.socket
```

And running an application similar to this one:

```js
/* Using built-in modules only here, since extra dependencies
 * would compromise testing too much.
 */
var http = require('http');

var responder = function (request, response) {
  response.writeHead(200, {'Content-Type': 'text/plain'});
  response.end('socket response for ' + request.url);
};

var server = http.createServer(responder);

/* Since this folder most likely is owned by root,
 * use `sudo` to run this script.
 */
server.listen('/var/run/www.socket');
```

When calling `http://127.0.0.1:2015/test` now, server replies with `socket response for /var/run/www.socket/test`.

# What should happen

When calling `http://127.0.0.1:2015/test` now, server should reply with `socket response for /test`.

# What I did

I added another block that behaves just like `without` would do, trimming the path of the socket file.
I'm not sure at the moment whether to put it above or below `without`, but I guess I've got it the wrong way around? I couldn't sleep this night. I didn't succeed with my switch yesterday because of several problems, one of which was exactly this issue.

### CODE REVIEW

**Comment by mholt** [comment]
Hey, nice work! I didn't think about that, so I'm glad you're fixing it.

The PR looks good. As far as whether it should go before or after the 'without' logic, try writing a test. Once this PR has a test we can be confident it is working as expected. The test can try a couple different scenarios (e.g. with 'without' logic and without 'without' logic -- phew, that was weird!) and if it passes we can merge it in! :+1:

Happy Easter, by the way. Hope you can sleep better tonight :smile:

**Comment by eiszfuchs** [comment]
Ahoy!

I've been able to make quite a mess, using *parts* of my previous attempts at writing a test for the socket functionality. This passes and covers my changes. :+1: 
I might add that for regular `http` proxies as well.

I see the build is failing at the linting of the `markdown` middleware. Is that my fault?

**Comment by mholt** [comment]
@eiszfuchs I added a new test to the CI script this weekend that checks `gofmt -s`, and this is the first time a push has failed on it; I didn't realize it wouldn't output anything useful! But it looks like proxy_test.go needs to be `go fmt`'ed:

```diff
< 	name string
---
> 	name    string
```

Line 368.

I'll work on getting more useful output for that test, sorry! Tests look great at a glance!

**Comment by eiszfuchs** [comment]
This is really weird, but I appreciate clean code.
Didn't know it would fail on some missing spaces. Hah!

Will squash and add a HTTP proxy test as well later (even though this will double the test code added, which doesn't sound *too* bad actually).

Wait. It's still failing on `go vet ./...`. What the heck? `caddy/config` and `server/server.go` seem to be broken.

**Comment by mholt** [comment]
@eiszfuchs It's `go vet` that's failing, and [I believe it's a bug in go vet on Go tip](https://github.com/golang/go/issues/14994). Notice that it succeeds on Go 1.6.

Anyway, your PR is now passing, so I will merge in as soon as you add a HTTP proxy test and squash commits. Then we'll make you a collaborator. :+1:

**Comment by eiszfuchs** [comment]
Yes, I noticed it wouldn't fail with 1.6 (since that's installed on my machine, too).
Additional test might take until the weekend. :briefcase:

**Comment by mholt** [comment]
In that case, is it alright if I merge your PR now? Just squash your commits first.

**Comment by eiszfuchs** [comment]
No, wait with that. That error has been in the code for a while and I'd feel more comfortable with that additional test before merging that into `master`. Safety first! :)
If I happen to find a clever solution to write the test efficiently without the whole lot of copy/paste insanity, I'll add it right away.

**Comment by mholt** [comment]
Okay, I'll wait. :smile: Just wanted to avoid possible merge conflicts. Thanks for being careful!

**Comment by eiszfuchs** [comment]
I have been rebasing before, I think I'll be able to do that magic trick again. :+1:

**Comment by eiszfuchs** [comment]
Ahoy! I made it!
Well, the tests got a lot more complicated, but if you think they're looking fine, that's all I need.

File: middleware/proxy/proxy_test.go [hunk 1]
```diff
@@ -238,6 +239,118 @@ func TestUnixSocketProxy(t *testing.T) {
...
+	res.Body.Close()
+	if err != nil {
+		return errors.New(fmt.Sprintf("Unable to read body: %v", err)), ""
+	}
+
+	return nil, fmt.Sprintf("%s", greeting)
+}
+
+func TestUnixSocketProxyPaths(t *testing.T) {
+	if runtime.GOOS == "windows" {
```
Comment (eiszfuchs) [inline middleware/proxy/proxy_test.go]: Oh wait! This only needs to happen for the socket test in line 336, doesn't it?

Comment (mholt) [inline middleware/proxy/proxy_test.go]: Yep, I think so!

File: middleware/proxy/proxy_test.go [hunk 2]
```diff
@@ -238,6 +239,118 @@ func TestUnixSocketProxy(t *testing.T) {
 	}
 }
 
+func GetHttpProxy(messageFormat string, prefix string) (*Proxy, *httptest.Server) {
+	ts := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, messageFormat, r.URL.String())
+	}))
+
+	ts.Start()
```
Comment (mholt) [inline middleware/proxy/proxy_test.go]: Since you're starting the server right away, you might be able to just use `httptest.NewServer()`

**Comment by eiszfuchs** [comment]
Uhm, yes, made changes according to comments. It's getting late and I'm getting nervous. :disappointed_relieved:

**Comment by mholt** [comment]
@eiszfuchs If you want to fix the lint warnings now, that'd be superb. (Unfortunately there's nothing we can do about the first two.) Then I think we're good to go! (don't forget to squash)

**Comment by eiszfuchs** [comment]
Will eat something, then sure, I love* fixing lint warnings!

* That's some Javascript developer right here

**Comment by eiszfuchs** [comment]
Lint warnings are gone! :tada:

**Comment by mholt** [comment]
Woo hoo! Excellent, thank you so much!

Merging now.

**Comment by mholt** [comment]
I've also added you as a collaborator, so feel free to participate in code reviews of other PRs, or submit others as you would like, you can even merge PRs if everything looks good (commits squashed, CI tests pass, etc)!

**Comment by eiszfuchs** [comment]
Oh! Thanks! Will see what I can do! :beers:

**Pull request merged**